contentgrid-appserver-query-engine-impl-jooq/src/test/java/com/contentgrid/appserver/query/engine/jooq/EventsDispatchTest.java

```diff
@@ -0,0 +1,334 @@
+                                .build())
+                        .build(),
+                PERMIT_ALWAYS
+        );
+
+        Mockito.verify(eventConsumer).dispatchUpdate(
+                eq(APPLICATION),
+                eq(ENTITY_A.getName()),
+                argThat(oldData -> {
+                    var attribute = oldData.getAttributeByName(ATTRIBUTE_A.getName());
+                    if (attribute.isPresent() && attribute.get() instanceof SimpleAttributeData<?> attr) {
+                        assertThat(attr.getValue()).isEqualTo(123L);
+                        return true;
+                    }
+                    return false;
+                }),
+                argThat(newData -> {
+                    var attribute = newData.getAttributeByName(ATTRIBUTE_A.getName());
+                    if (attribute.isPresent() && attribute.get() instanceof SimpleAttributeData<?> attr) {
+                        assertThat(attr.getValue()).isEqualTo(456L);
+                        return true;
+                    }
+                    return false;
+                })
+        );
+
+        // Transaction completed → it's updated in db
+        var updated = queryEngine.findById(APPLICATION, created.getIdentity().toRequest(), PERMIT_ALWAYS).orElseThrow();
+        var updatedAttr = (SimpleAttributeData<?>) updated.getAttributeByName(ATTRIBUTE_A.getName()).orElseThrow();
+        assertThat(updatedAttr.getValue()).isEqualTo(456L);
+    }
+
+    @Test
+    void verifyDelete_happy() {
+        var created = queryEngine.create(
+                APPLICATION,
+                EntityCreateData.builder()
+                        .entityName(ENTITY_A.getName())
+                        .attribute(SimpleAttributeData.builder()
+                                .name(ATTRIBUTE_A.getName())
+                                .value(123)
+                                .build())
+                        .build(),
+                PERMIT_ALWAYS
+        );
+
+        queryEngine.delete(
+                APPLICATION,
+                EntityRequest.forEntity(created.getName(), created.getId()),
+                PERMIT_ALWAYS
+        );
+
+        Mockito.verify(eventConsumer).dispatchDelete(
+                eq(APPLICATION),
+                eq(ENTITY_A.getName()),
+                argThat(oldData -> {
+                    var attribute = oldData.getAttributeByName(ATTRIBUTE_A.getName());
+                    if (attribute.isPresent() && attribute.get() instanceof SimpleAttributeData<?> attr) {
+                        assertThat(attr.getValue()).isEqualTo(123L);
+                        return true;
+                    }
+                    return false;
+                })
+        );
+
+        // Transaction completed → it's deleted in db
+        var missing = queryEngine.findById(APPLICATION, created.getIdentity().toRequest(), PERMIT_ALWAYS);
+        assertThat(missing).isEmpty();
+    }
+
+    @Test
+    void verifyCreate_unhappy() {
```

Comment:
Do you want to test here that the transaction is rolled back when the event is failing to be emitted?

In that case, I would create a specific exception for that, just in the test class here. That way, you are sure that the exception emitted by the event here is handled.

Also, maybe rename the test to `create_eventThrows()`.

There is also another unhappy case when the create just fails (e.g. because of a duplicate key error during insert). You forgot to cover those to check that no events are emitted in that case.